### PR TITLE
add model scaling section

### PR DESCRIPTION
**Pull request by lvwerra**
# What does this PR do?

This PR adds a practical scaling guide to the documentation. When starting to work with large transformer models one of the most common issues is running out of GPU memory. There are several easy to implement strategies to counteract this. This guide shows how to use gradient accumulation, gradient checkpointing, mixed precision training, and choice of optimizer to decrease the memory footprint. It shows how to do this with both the `Trainer` and `accelerate` in a hands-on fashion.

## What does this PR not do?
This PR does **not** replace the in-depth explanations at ["Model Parallelism"](https://huggingface.co/docs/transformers/parallelism) and ["Performance and Scalability: How To Fit a Bigger Model and Train It Faster"](https://huggingface.co/docs/transformers/performance). Rather the proposed guide should be an easy entry point for new users whereas these other sections are for more advanced users.

## TensorFlow?
If this guide resonates well it would make sense to extend it with a TensorFlow/Keras section (similar to the `accelerate` section). Maybe TensorFlow boy (aka @Rocketknight1) has some ideas?

### CODE REVIEW

**Comment by Rocketknight1** [comment]
We could definitely make a TF/Keras section! I believe Keras doesn't support gradient accumulation or gradient checkpointing out of the box, though, but mixed precision, Adafactor, etc. are all very doable.

File: docs/source/scaling.mdx [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+default_args = {
+    "output_dir": "tmp",
+    "evaluation_strategy": "steps",
+    "num_train_epochs": 1,
+    "log_level": "error",
+    "report_to": "none",
+}
+```
+
+> Note: In order to properly clear the memory after experiments we need restart the Python kernel between experiments. Run all steps above and then just one of the experiments below.
```
Comment (sgugger) [inline docs/source/scaling.mdx:131]: Use the <Tip> syntax here for better rendering.

File: docs/source/scaling.mdx [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+
+## Vanilla Training
+
+As a first experiment we will use the `Trainer` and train the model without any further modifications and a batch size of 4:
```
Comment (sgugger) [inline docs/source/scaling.mdx:135]:
```suggestion
As a first experiment we will use the [`Trainer`] and train the model without any further modifications and a batch size of 4:
```

File: docs/source/scaling.mdx [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+```py
+training_args = TrainingArguments(per_device_train_batch_size=4, **default_args)
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 57.82
+    Samples/second: 8.86
+    GPU memory occupied: 14949 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:146]: Use the doctest syntax or put the result in a code block.

File: docs/source/scaling.mdx [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+
+We see that already a relatively small batch size almost fills up our GPU's entire memory. However, a larger batch size can often result in faster model convergence or better end performance. So ideally we want to tune the batch size to our model's needs and not to the GPU limitations. A simple trick to effectively train larger batch size is gradient accumulation.
+
+## Gradient Accumulation
+
+The idea behind gradient accumulation is to instead of calculating the gradients for the whole batch at once to do it in smaller steps. The way we do that is to calculate the gradients iteratively in smaller batches by doing a forward and backward pass through the model and accumulating the gradients in the process. When enough gradients are accumulated we run the model's optimization step. This way we can save GPU memory but in turn we are adding more forward and backward passes which can slow down the training.
+
+We can use gradient accumulation in the `Trainer` by simply adding the `gradient_accumulation_steps` argument to `TrainingArguments`. Let's see how it impacts the models memory footprint:
```
Comment (sgugger) [inline docs/source/scaling.mdx:155]:
```suggestion
We can use gradient accumulation in the [`Trainer`] by simply adding the `gradient_accumulation_steps` argument to [`TrainingArguments`]. Let's see how it impacts the models memory footprint:
```

File: docs/source/scaling.mdx [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+    Time: 57.82
+    Samples/second: 8.86
+    GPU memory occupied: 14949 MB.
+
+
+We see that already a relatively small batch size almost fills up our GPU's entire memory. However, a larger batch size can often result in faster model convergence or better end performance. So ideally we want to tune the batch size to our model's needs and not to the GPU limitations. A simple trick to effectively train larger batch size is gradient accumulation.
+
+## Gradient Accumulation
+
+The idea behind gradient accumulation is to instead of calculating the gradients for the whole batch at once to do it in smaller steps. The way we do that is to calculate the gradients iteratively in smaller batches by doing a forward and backward pass through the model and accumulating the gradients in the process. When enough gradients are accumulated we run the model's optimization step. This way we can save GPU memory but in turn we are adding more forward and backward passes which can slow down the training.
```
Comment (sgugger) [inline docs/source/scaling.mdx:153]: It's not obvious from the intro that this is just a way to train at a bigger batch size while not going OOM.

File: docs/source/scaling.mdx [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+training_args = TrainingArguments(per_device_train_batch_size=1, gradient_accumulation_steps=4, **default_args)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 66.03
+    Samples/second: 7.75
+    GPU memory occupied: 8681 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:167]: Same remark as above.

File: docs/source/scaling.mdx [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+
+Next we have a look at another trick to save a little bit more GPU memory called gradient checkpointing.
+
+## Gradient Checkpointing
+
+Even when we set the batch size to 1 and use gradient accumulation we can still run out of memory when working with large models. In order to compute the gradients during the backward pass all activations from the forward pass are normally saved. This can create a big memory overhead. Alternatively, one could forget all activations during the forward pass and recompute them on demand during the backward pass. This would however add a significant computational overhead and slow down training.
+
+Gradient checkpointing strikes a compromise between the two approaches and saves strategically selected activations throughout the computational graph so only a fraction of the activations need to be re-computed for the gradients. See [this great article](https://medium.com/tensorflow/fitting-larger-networks-into-memory-583e3c758ff9) explaining the ideas behind gradient checkpointing. 
+
+To enable gradient checkpointing in the `Trainer` we only need ot pass it as a flag to the `TrainingArguments`. Everything else is handled under the hood:
```
Comment (sgugger) [inline docs/source/scaling.mdx:180]:
```suggestion
To enable gradient checkpointing in the [`Trainer`] we only need ot pass it as a flag to the [`TrainingArguments`]. Everything else is handled under the hood:
```

File: docs/source/scaling.mdx [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 85.47
+    Samples/second: 5.99
+    GPU memory occupied: 6775 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:194]: Same remark as above.

File: docs/source/scaling.mdx [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+
+
+We can see that this saved some more memory but at the same time training became a bit slower. A general rule of thumb is that gradient checkpointing slows down training by about 20%. Let's have a look at another method with which we can regain some speed and also reduce the memory footprint: mixed precision training.
+
+## FP16 Training
+
+The idea of mixed precision training is that no all variables need to be stored in full (32-bit) floating point precision. If we can reduce the precision the variales and their computations are faster and use less memory. The main advantage comes from saving the activations in half (16-bit) precision. Although the gradients are also computed in half precision they are converted back to full precision for the optimization step so no memory is saved here. For this reason this approach is also called mixed precision training. Enabling mixed precision training is also just a matter of setting the `fp16` flag to `True`:
```
Comment (sgugger) [inline docs/source/scaling.mdx:201]: No, fp16 actually results in more memory usage in most cases when batch size is small and the model is big, because there is 1.5 x the size of the model on the GPU (one copy in FP32 and one copy in FP16).

This is a common source of confusion for users.

File: docs/source/scaling.mdx [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+print_summary(result)
+```
+
+    Time: 27.46
+    Samples/second: 18.64
+    Loss: 0.02
+    GPU memory occupied: 13939 MB.
+
+
+We can see that this is almost twice as fast as the vanilla training while also requiring less GPU memory. Let's add it to the mix of the previous methods:
```
Comment (sgugger) [inline docs/source/scaling.mdx:217]: I would remove the "requiring less GPU memory" as it's only a few MBs.

File: docs/source/scaling.mdx [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 50.76
+    Samples/second: 10.09
+    GPU memory occupied: 7275 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:236]: Same remark as above.

File: docs/source/scaling.mdx [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+training_args = TrainingArguments(per_device_train_batch_size=4, adafactor=True, **default_args)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 64.31
+    Samples/second: 7.96
+    GPU memory occupied: 12295 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:260]: Same remark as above.

File: docs/source/scaling.mdx [hunk 13]
```diff
@@ -0,0 +1,446 @@
...
+)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 56.54
+    Samples/second: 9.06
+    GPU memory occupied: 4847 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:283]: Same remark as above.

File: docs/source/scaling.mdx [hunk 14]
```diff
@@ -0,0 +1,446 @@
...
+
+
+We went from 15 GB memory usage to 5 GB - a 3x improvement while maintaining the throughput! However, as mentioned before, the convergence of Adafactor can be worse than Adam. There is an alternative to Adafactor called 8-bit Adam that takes a slightly different approach.
+
+### 8-bit Adam
+
+Instead of aggregating optimizer states like Adafactor, 8-bit Adam keeps the full state and quantizes it. Quantization means that it stores the state with lower precision and dequantizes it only for the optimization. This is similar to the idea behind FP16 training where using variables with lower precision saves memory. 
+
+In contrast to the previous approaches is this one not integrated into the `Trainer` as a simple flag. We need to install the 8-bit optimizer and then pass it as a custom optimizer to the `Trainer`. Follow the installation guide in the Github [repo](https://github.com/facebookresearch/bitsandbytes) to install the `bitsandbytes` library that implements the 8-bit Adam optimizer.
```
Comment (sgugger) [inline docs/source/scaling.mdx:292]:
```suggestion
In contrast to the previous approaches is this one not integrated into the [`Trainer`] as a simple flag. We need to install the 8-bit optimizer and then pass it as a custom optimizer to the [`Trainer`]. Follow the installation guide in the Github [repo](https://github.com/facebookresearch/bitsandbytes) to install the `bitsandbytes` library that implements the 8-bit Adam optimizer.
```
+ link to the bitsandbytes GitHub.

File: docs/source/scaling.mdx [hunk 15]
```diff
@@ -0,0 +1,446 @@
...
+We can now pass the custom optimizer as an argument to the `Trainer`:
+```py
+trainer = Trainer(model=model, args=training_args, train_dataset=ds, optimizers=(adam_bnb_optim, None))
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 55.95
+    Samples/second: 9.15
+    GPU memory occupied: 13085 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:333]: Same remark as above.

File: docs/source/scaling.mdx [hunk 16]
```diff
@@ -0,0 +1,446 @@
...
+        "weight_decay": 0.0,
+    },
+]
+
+optimizer_kwargs = {
+    "betas": (training_args.adam_beta1, training_args.adam_beta2),
+    "eps": training_args.adam_epsilon,
+}
+optimizer_kwargs["lr"] = training_args.learning_rate
+adam_bnb_optim = bnb.optim.Adam8bit(optimizer_grouped_parameters, **optimizer_kwargs)
```
Comment (sgugger) [inline docs/source/scaling.mdx:321]: Can we just pass the kwargs when calling `Adam8bit`? I don't see why we store them in a dict.

File: docs/source/scaling.mdx [hunk 17]
```diff
@@ -0,0 +1,446 @@
...
+)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds, optimizers=(adam_bnb_optim, None))
+result = trainer.train()
+print_summary(result)
+```
+
+    Time: 49.46
+    Samples/second: 10.35
+    GPU memory occupied: 5363 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:354]: Same remark as above.

File: docs/source/scaling.mdx [hunk 18]
```diff
@@ -0,0 +1,446 @@
...
+
+
+Again, we get about a 3x memory improvement and even slightly higher throughput as using Adafactor. So we have seen how we can optimize the memory footprint of large models. The following plot summarizes all our experiments:
+    
+![png](https://huggingface.co/datasets/lvwerra/repo-images/raw/main/gpu-memory-savings.png)
+
+## Using `accelerate`
```
Comment (sgugger) [inline docs/source/scaling.mdx:361]:
```suggestion
## Using 🤗 Accelerate
```

File: docs/source/scaling.mdx [hunk 19]
```diff
@@ -0,0 +1,446 @@
...
+
+So far we have used the `Trainer` to run the experiments but a more flexible alternative to that approach is to use [`accelerate`](https://huggingface.co/docs/accelerate/). With `accelerate` you have full control over the training loop and can essentially write the loop in pure PyTorch with some minor modifications. In turn it allows you to easily scale across different infrastructures such as CPUs, GPUs, TPUs, or distributed multi-GPU setups without chaning any code. Let's see what it takes to implement all of the above tweaks in accelerate. We can still use the `TrainingArguments` to wrap the training settings:
```
Comment (sgugger) [inline docs/source/scaling.mdx:363]:
```suggestion
So far we have used the [`Trainer`] to run the experiments but a more flexible alternative to that approach is to use [`accelerate`](https://huggingface.co/docs/accelerate/). With 🤗 Accelerate you have full control over the training loop and can essentially write the loop in pure PyTorch with some minor modifications. In turn it allows you to easily scale across different infrastructures such as CPUs, GPUs, TPUs, or distributed multi-GPU setups without changing any code. Let's see what it takes to implement all of the above tweaks in accelerate. We can still use the [`TrainingArguments`] to wrap the training settings:
```

File: docs/source/scaling.mdx [hunk 20]
```diff
@@ -0,0 +1,446 @@
...
+training_args = TrainingArguments(
+    per_device_train_batch_size=1,
+    gradient_accumulation_steps=4,
+    gradient_checkpointing=True,
+    fp16=True,
+    **default_args,
+)
+```
+
+The full example training loop with `accelerate` is only a handful of lines of code long:
```
Comment (sgugger) [inline docs/source/scaling.mdx:376]:
```suggestion
The full example training loop with 🤗 Accelerate is only a handful of lines of code long: 
```

File: docs/source/scaling.mdx [hunk 21]
```diff
@@ -0,0 +1,446 @@
...
+for step, batch in enumerate(dataloader, start=1):
+    loss = model(**batch).loss
+    loss = loss / training_args.gradient_accumulation_steps
+    accelerator.backward(loss)
+    if step % training_args.gradient_accumulation_steps == 0:
+        optimizer.step()
+        optimizer.zero_grad()
+```
+
+First we wrap the dataset in a `DataLoader`. Then we can enable gradient checkpointing by calling the model's `gradient_checkpointing_enable` method. When we initialize the `Accelerator` we can specifiy if we want to use mixed precision training and it will take care of it for us in the `prepare` call. During the `prepare` call the dataloader will also be distributed across workers should we use multiple GPUs. We use the same 8-bit optimizer from the earlier experiments.
```
Comment (sgugger) [inline docs/source/scaling.mdx:401]: Linking to the doc of all those objects would be great!

File: docs/source/scaling.mdx [hunk 22]
```diff
@@ -0,0 +1,446 @@
...
+
+Finally, we can write the main training loop. Note that the `backward` call is handled by accelerate. We can also see how gradient accumulation works: we normalize the loss so we get the average at the end of accumulation and once we have enough steps we run the optimization. Now the question is: does this use the same amount of memory as the previous steps? Let's check:
+
+
+```py
+print_gpu_utilization()
+```
+
+    GPU memory occupied: 5363 MB.
```
Comment (sgugger) [inline docs/source/scaling.mdx:410]: Use the doctest syntax here.

File: docs/source/scaling.mdx [hunk 23]
```diff
@@ -0,0 +1,446 @@
...
+
+
+Indeed it does. Implementing these optimization techniques with `accelerate` only takes a handful of lines of code and comes with the benefit of more flexiblity in the training loop.
```
Comment (sgugger) [inline docs/source/scaling.mdx:413]:
```suggestion
Indeed it does. Implementing these optimization techniques with 🤗 Accelerate only takes a handful of lines of code and comes with the benefit of more flexiblity in the training loop.
```

File: docs/source/scaling.mdx [hunk 24]
```diff
@@ -0,0 +1,446 @@
...
+
+## Multi-GPU Training
+
+If your model fits on a single GPU scaling to many GPUs can be achieved fairly easily with data parallelism. The idea is very similar to gradient accumulation with the distinction that instead of running the forward and backward passes during the accumulation in sequence on a single machine they are performed in parallel on multiple machines. So each GPU gets a small batch, runs the forward and backward passes and then the gradients from all machines are aggregated and the model is optimized. You can combine this with all the methods we described before. For example, if you have 4 GPUs and use `per_device_train_batch_size=12` and `gradient_accumulation_steps=3` you will have an effective batch size of `4*12*3=144`. 
+
+The `Trainer` allows for distributed training and if you execute your `Trainer` training script on a machine with multiple GPUs it will automatically utilize all of them, hence the name `per_device_train_batch_size`. In `accelerate` you can configure the infrastructure setup with the following command:
```
Comment (sgugger) [inline docs/source/scaling.mdx:434]:
```suggestion
The [`Trainer`] allows for distributed training and if you execute your [`Trainer`] training script on a machine with multiple GPUs it will automatically utilize all of them, hence the name `per_device_train_batch_size`. In 🤗 Accelerate you can configure the infrastructure setup with the following command:
```

File: docs/source/scaling.mdx [hunk 25]
```diff
@@ -0,0 +1,446 @@
...
+
+```bash
+accelerate config
+```
+
+Until now we have opperated under the assumption that we can fit the model onto a single GPU without or with the introduced tricks . But what if this is not possible? We still have a few tricks up our sleeves!
+
+## What if my model still does not fit?
+
+If the model does not fit on a single GPU with all the mentioned tricks there are still more methods we can apply although life starts to get a bit more complicated. This usually involves some form of pipeline or tensor parallelism where the model itself is distributed across several GPUs. One can also make use of DeepSpeed which implements some of these parallelism strategies along with some more optimization to reduce the memory footprint such as partitioning the optimizer states. You can read more about this in the ["Model Parallelism" section](https://huggingface.co/docs/transformers/parallelism).
```
Comment (sgugger) [inline docs/source/scaling.mdx:444]:
```suggestion
If the model does not fit on a single GPU with all the mentioned tricks there are still more methods we can apply although life starts to get a bit more complicated. This usually involves some form of pipeline or tensor parallelism where the model itself is distributed across several GPUs. One can also make use of DeepSpeed which implements some of these parallelism strategies along with some more optimization to reduce the memory footprint such as partitioning the optimizer states. You can read more about this in the ["Model Parallelism" section](parallelism). 
```

File: docs/source/scaling.mdx [hunk 26]
```diff
@@ -0,0 +1,446 @@
...
+
+## What if my model still does not fit?
+
+If the model does not fit on a single GPU with all the mentioned tricks there are still more methods we can apply although life starts to get a bit more complicated. This usually involves some form of pipeline or tensor parallelism where the model itself is distributed across several GPUs. One can also make use of DeepSpeed which implements some of these parallelism strategies along with some more optimization to reduce the memory footprint such as partitioning the optimizer states. You can read more about this in the ["Model Parallelism" section](https://huggingface.co/docs/transformers/parallelism). 
+
+This concludes this guide for scaling the training of large models. There is a lot more going on under the hood than we looked at and the ["Performance and Scalability: How To Fit a Bigger Model and Train It Faster" section](https://huggingface.co/docs/transformers/performance) goes into much more detail about the introduced optimization techniques we introduced.
```
Comment (sgugger) [inline docs/source/scaling.mdx:446]:
```suggestion
This concludes this guide for scaling the training of large models. There is a lot more going on under the hood than we looked at and the ["Performance and Scalability: How To Fit a Bigger Model and Train It Faster" section](performance) goes into much more detail about the introduced optimization techniques we introduced.
```

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this, this is a great addition!

The way you show the outputs is inconsistent with the rest of the doc. It should either be with the doctest syntax:
```
>>> instruction
result 
```
or in two separate code blocks.

File: docs/source/scaling.mdx [hunk 27]
```diff
@@ -0,0 +1,446 @@
...
+
+
+```py
+from pynvml import *
+import numpy as np
+
+from datasets import Dataset
+from transformers import AutoTokenizer, AutoModelForSequenceClassification, TrainingArguments, Trainer, logging
+
+logging.set_verbosity_error()
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply docs/source/scaling.mdx:34]: Only the first time they appear or every time? E.g. the `Trainer` is used all over and in order to make them individually copy-pastable I'd have to import it in every cell, right?

[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/scaling.mdx:34]: Only the first time is fine!

File: docs/source/scaling.mdx (hunk 5 above)
Comment (lvwerra) [reply docs/source/scaling.mdx:153]: I modified the last sentence to reflect that better.

File: docs/source/scaling.mdx (hunk 9 above)
Comment (lvwerra) [reply docs/source/scaling.mdx:201]: including me :) i fixed the text.

File: docs/source/scaling.mdx (hunk 16 above)
Comment (lvwerra) [reply docs/source/scaling.mdx:321]: Good point, a relic from copy-pasting.

**Comment by lvwerra** [comment]
Hi @sgugger I integrated your feedback, let me know if this is what you had in mind.

File: docs/source/scaling.mdx [hunk 28]
```diff
@@ -0,0 +1,454 @@
...
+
+Let's verify that we start with a free GPU memory:
+
+```py
+>>> print_gpu_utilization()
+GPU memory occupied: 0 MB.
+```
+
+That looks good: the GPU memory is not occupied as we would expect before we load any models. If that's not the case on your machine make sure to stop all processes that are using GPU memory.
+
```
Comment (stas00) [inline docs/source/scaling.mdx:70]: at this point I typically do `torch.ones((1, 1)).to(device_id)` because cuda kernels take up 1-2GB of each card.

https://github.com/stas00/ipyexperiments/blob/39c9b454e89e53b74c74dcb579c12ecf3d2161b9/ipyexperiments/utils/mem.py#L21-L28

Because not all of card's memory is available to the user, but once it's loaded the user knows how much memory is really available.

**Review by stas00 — COMMENTED** [review]
Thanks for working on it, @lvwerra. We need more of similar content.

That's said I think a better solution would be to rewrite the existing documents and not add a new one that has more than 50% overlap with the existing doc, which will make it more confusing to the user and more difficult to maintain for us, as many things will now be duplicated if this PR is merged as is.

Indeed the original performance doc has been grown into scalability+performance and parallelism is no longer parallelism but actually it's the scalability doc, that's specifically there to deal with OOM.

I think we need 2 docs:

1. scalability
2. performance

and move/combine this PR's doc with the scalability sections of the the current performance doc - removing the huge overlap that this PR proposes.

and for the performance doc leave only software/hardware bits that talk about speeding things up, and minimally talking about memory usage.

Bottom line, IMHO it'd be good to brainstorm how to do that in the most harmonious way before creating a new document.

**Comment by sgugger** [comment]
In the case of documentation, I don't think repetition is a bad thing since users rarely look at every page in a row. Making sure the docs are all linked to each other for further reading is important. The two doc pages reach different target audience IMO and present the techniques in different ways, so I think having both of them is good.

**Comment by stas00** [comment]
Repetition is a bad thing in this situation, IMHO:

1. as it'd make it more difficult to point users to the right doc. For example @LysandreJik has been pointing users to the performance doc when they have difficulties with getting their model to run memory-wise. Now you will have 2 competing docs, which one should he point users to? It shouldn't take more than a few secs for a maintainer to find where to point the user to and not needing to reread each doc and contemplate is it A or B?
2. how do you maintain 2 overlapping documents?
3. some users do read docs and now they will need to read 2 docs with confusing overlaps

But the big question is this why make things more complicated and messy when the whole performance/scalability document is due to a revamp anyway. It's far from great at the moment. I propose to discuss how we best serve both users and ourselves by improving the performance/scalability doc structure/layout. I'm not proposing to dropping anything from the current PR but to integrate rather than creating a diverging doc.

**Comment by stas00** [comment]
update: @lvwerra and I will discuss a revamp/integration on Fri.

**Comment by lvwerra** [comment]
Hi there, I updated the doc and this is ready for another review:
- use @stas00's suggestion to initialize the measurement by loading a small tensor to the GPU
- added a note about the embedding issue with BnB optimizer

I disusssed with @stas00 that we could start the revamp by merging this PR and then move things around as we start adding the other documents in #15213.

**Review by stas00 — APPROVED** [review]
Looks great, @lvwerra 

I just wonder whether for now we should append this to the end of `performance.mdx`? Since we don't intend to have a `scaling.mdx` in the revamped docs - at least not in the initial layout.

otherwise if people start linking to `scaling.mdx` and then we wipe it out - we will need to deal with additional redirects or end up with broken links.

or at the very least putting a big fat disclaimer that this doc is temporary and will be removed soon. do not link to it.

**Comment by lvwerra** [comment]
Good point, I merged `scaling.mdx` and `performance.mdx`. The document starts with the new practical guide and I added the existing material to a section called `Further discussions`. There is still a bit of redundancy between the two parts but I think we can refactor it when we flesh out the other parts.

@sgugger are you happy with the current state to be merged?

File: docs/source/_toctree.yml [hunk 29]
```diff
@@ -25,6 +25,7 @@
     title: Summary of the tokenizers
   - local: multilingual
     title: Multi-lingual models
+
```
Comment (sgugger) [inline docs/source/_toctree.yml:28]: Not sure this new line is intentional. It shouldn't be needed.

File: docs/source/performance.mdx [hunk 30]
```diff
@@ -15,10 +15,466 @@ limitations under the License.
 -->
 
 # Performance and Scalability: How To Fit a Bigger Model and Train It Faster
+> _Or how to escape the dreaded "RuntimeError: CUDA error: out of memory" error._
```
Comment (sgugger) [inline docs/source/performance.mdx:18]: A new line here would be nice for readability.

File: docs/source/performance.mdx [hunk 31]
```diff
@@ -15,10 +15,466 @@ limitations under the License.
...
+result = trainer.train()
+print_summary(result)
+```
+```
+Time: 66.03
+Samples/second: 7.75
+GPU memory occupied: 8681 MB.
+```
+
+We can see that the memory footprint was dramatically reduced at the cost of being only slightly slower than the vanilla run. Of course, this would change as you increase the number of accumulation steps. In general you would want to max out the GPU usage as much as possible. So in our case, the batch_size of 4 was already pretty close to the GPU's limit. If we wanted to train with a batch size of 64 we should not use `per_device_train_batch_size=1` and `gradient_accumulation_steps=64` but instead `per_device_train_batch_size=4` and `gradient_accumulation_steps=16` which has the same effective batch size while making better use of the available GPU resources.
```
Comment (sgugger) [inline docs/source/performance.mdx:190]: Very nice remark!

File: docs/source/performance.mdx [hunk 32]
```diff
@@ -15,10 +15,466 @@ limitations under the License.
...
+The idea of mixed precision training is that no all variables need to be stored in full (32-bit) floating point precision. If we can reduce the precision the variales and their computations are faster. The main advantage comes from saving the activations in half (16-bit) precision. Although the gradients are also computed in half precision they are converted back to full precision for the optimization step so no memory is saved here. Since the model is present on the GPU in both 16-bit and 32-bit precision this can use more GPU memory (1.5x the original model is on the GPU), especially for small batch sizes. Since some computations are performed in full and some in half precision this approach is also called mixed precision training. Enabling mixed precision training is also just a matter of setting the `fp16` flag to `True`:
+
+```py
+training_args = TrainingArguments(per_device_train_batch_size=4, fp16=True, **default_args)
+
+trainer = Trainer(model=model, args=training_args, train_dataset=ds)
+result = trainer.train()
+print_summary(result)
+```
+```
```
Comment (sgugger) [inline docs/source/performance.mdx:230]: Adding a new line here for readability would be nice.

File: docs/source/performance.mdx [hunk 33]
```diff
@@ -15,10 +15,466 @@ limitations under the License.
...
+
+The most common optimizer used to train transformer model is Adam or AdamW (Adam with weight decay). Adam achieves good convergence by storing the rolling average of the previous gradients which, however, adds an additional memory footprint of the order of the number of model parameters. One remedy to this is to use an alternative optimizer such as Adafactor.
+
+### Adafactor
+
+Instead of keeping the rolling average for each element in the weight matrices Adafactor only stores aggregated information (row- and column-wise sums of the rolling averages) which reduces the footprint considerably. One downside of Adafactor is that in some instances convergence can be slower than Adam's so some experimentation is advised here. We can use Adafactor simply by setting `adafactor=True`:
+
+
+```py
+training_args = TrainingArguments(per_device_train_batch_size=4, adafactor=True, **default_args)
```
Comment (sgugger) [inline docs/source/performance.mdx:270]: I think this way is deprecated and that one has to pass optim-"adafactor" nowadays. @stas00 can confirm.

File: docs/source/performance.mdx [hunk 34]
```diff
@@ -15,10 +15,466 @@ limitations under the License.
...
+We can see that this saves a few more GB on the GPU. Let's see how it looks when we add it to the other methods we introduced earlier:
+
+
+```py
+training_args = TrainingArguments(
+    per_device_train_batch_size=1,
+    gradient_accumulation_steps=4,
+    gradient_checkpointing=True,
+    fp16=True,
+    adafactor=True,
```
Comment (sgugger) [inline docs/source/performance.mdx:291]: Same here.

**Review by sgugger — APPROVED** [review]
LGTM, thanks for iterating on this!

File: docs/source/_toctree.yml (hunk 29 above)
Comment (lvwerra) [reply docs/source/_toctree.yml:28]: Not sure where that came from - deleted.

File: docs/source/performance.mdx (hunk 33 above)
Comment (lvwerra) [reply docs/source/performance.mdx:270]: Indeed - fixed.

File: docs/source/performance.mdx (hunk 32 above)
Comment (lvwerra) [reply docs/source/performance.mdx:230]: added a newline between every input/output block in the document

File: docs/source/performance.mdx (hunk 30 above)
Comment (lvwerra) [reply docs/source/performance.mdx:18]: added

**Comment by stas00** [comment]
Thanks, @lvwerra 

1. Please revert all the header level changes in the existing doc - the menu on the right only handles ## and ### and your proposed change will practically eliminate most of the menu entries as you added an additional level

2. We discussed adding this PR's content at the end of the existing document as at the moment it's the extra and not the main document as it was laid out. 95% of this PR's material will be spread out over the new documents, and the current performance doc will remain here as a reference that the more specific docs will point to for indepth information. (e.g. mixed precision section we aren't going to repeat the full discussion about why and how in each sub-document and will point to the original instead). 

Does it make sense? If not, and we need to sync on the big picture - let's discuss it then first.

**Comment by lvwerra** [comment]
Hi @stas00 

Sure, I'll revert the headers, I was not aware of that limitation.

Regarding the order of the content: I know that we will change that document radically in later PRs but I thought since this is gonna be public and maybe here for a few weeks it would be nicer to have the practical guide at the beginning and the theoretical part at the end. If you feel strongly about this I can change it back.

**Comment by stas00** [comment]
> Hi @stas00
> 
> Sure, I'll revert the headers, I was not aware of that limitation.

In fact I asked when the new design was added to support more than 3-levels but nothing came out of that. i.e. the menu is missing already a bunch of entries where there is a deeper nesting like the deepspeed doc. :(

> Regarding the order of the content: I know that we will change that document radically in later PRs but I thought since this is gonna be public and maybe here for a few weeks it would be nicer to have the practical guide at the beginning and the theoretical part at the end. If you feel strongly about this I can change it back.

Sure, that's fine. I think what I will do instead in the new PR is to move this original document out to a new document.

**Comment by LysandreJik** [comment]
Thanks a lot for working on this important part of the docs!
